### ISSUE TITLE
OL 5.0.0 - ScaleLine slightly incorrect for EPSG:4326 with non-degree display units

### ISSUE THREAD

**Issue by cduplichien** [openlayers/openlayers#8370]
It's somewhat difficult to show this issue, as it only becomes apparent if you have both a scale line and a measure widget, and it's only off about 10% at our test latitude. In OL 4.6+ there have been several fixes to the ScaleLine control related to this, but minor discrepancies still exist if you have a projection in EPSG:4326 and ask the ScaleLine control to display in 'us' or some other non-degrees units.
This appears to be because the point resolution is calculated using the display units rather than the projection units, in ScaleLine.js 201-211:
```
  const projection = viewState.projection;
  const units = this.getUnits();
  const pointResolutionUnits = units == Units.DEGREES ?
    ProjUnits.DEGREES :
    ProjUnits.METERS;
  let pointResolution =
      getPointResolution(projection, viewState.resolution, center, pointResolutionUnits);
  if (projection.getUnits() != ProjUnits.DEGREES && projection.getMetersPerUnit()
    && pointResolutionUnits == ProjUnits.METERS) {
    pointResolution *= projection.getMetersPerUnit();
  }
```
In our case this means that the getPointResolution method is invoked using a unit type of 'us', even though the projection is in degrees. The documentation for getPointResolution mentions:
> For projections with degrees as the unit this will simply return the
>  provided resolution. For other projections the point resolution is
>  by default estimated by transforming the 'point' pixel to EPSG:4326,
>  measuring its width and height on the normal sphere,
>  and taking the average of the width and height.

The latter case appears to be taking effect: even though our projection is EPSG:4326 and the unit should be degrees, the 'us' display units are being passed in as an option override, causing the height-width average to be used instead. This gives us an incorrect scale line width.

It seems like it may be safer to always first compute the target width of the scale line using the projection units without an override, and then convert that width to the requested display units for the text output.

**Comment by ahocevar** [comment]
Thanks for the report. Can you provide a pull request?

**Comment by cduplichien** [comment]
It'll be a few weeks before I'm able to circle back to this issue.

**Comment by cduplichien** [comment]
I've looked back at this issue today. My previously suggested approach isn't a good solution.
The getPointResolution method was mostly doing what ScaleLine needed it to, even for EPSG-4326 with the non-degrees opt_units argument. Our only contention was with the returned "average" resolution between height and width resolutions, which caused the ScaleLine to not line up with our measure tool result if we were to measure the ScaleLine.
Internally, we opted to modify getPointResolution to accept a "dimension" argument from ScaleLine to allow it to return the "width" resolution instead of the "average" resolution, since the scale line by default is displayed horizontally on screen.
However, I don't think that the reported issue is actually a bug because even correcting for the width, the scale line is only usable for horizontal measurements and is misleading otherwise. Due to this, I believe EPSG:4326 with non-degree display units is just a misuse of the ScaleLine widget, and (especially in the presence of a map rotation control) the ScaleLine showing an average seems as reasonable as any alternative.
